experimental/serde_encoding/src/lib.rs

```diff
@@ -0,0 +1,54 @@
+use oak::{
+    handle::HandleVisit,
+    io::{Decodable, Encodable, Message},
+    OakError,
+};
+use serde::{Deserialize, Serialize};
+
+pub struct Bincoded<T>(T);
+
+impl<T: Clone + Serialize + HandleVisit> Encodable for Bincoded<T> {
+    fn encode(&self) -> Result<Message, OakError> {
+        Ok(Message {
+            bytes: bincode::serialize(&self.0).map_err(|_| OakError::ProtobufEncodeError(None))?,
+            handles: oak::handle::extract_handles(&mut self.0.clone()),
+        })
+    }
+}
+
+impl<'de, T: Clone + Deserialize<'de> + HandleVisit> Decodable for Bincoded<T> {
+    fn decode(message: &Message) -> Result<Self, OakError> {
+        // Yikes! Need to fix lifetimes
+        let clone = Box::leak(Box::new(message.clone()));
+        let mut decoded =
+            bincode::deserialize(&clone.bytes).map_err(|_| OakError::ProtobufDecodeError(None))?;
+        oak::handle::inject_handles(&mut decoded, &message.handles)?;
+        Ok(Bincoded(decoded))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use oak::io::{Encodable, Receiver};
+
+    #[derive(Clone, Serialize, Deserialize, HandleVisit, Debug, PartialEq)]
+    struct Test {
+        a: String,
+        b: Receiver<()>,
```

Comment:
Does it work with multiple fields having handles? That's the trick case (though I think your code should work)

Reply:
I think this is already covered in generic `HandleVisit` tests such as https://github.com/project-oak/oak/blob/main/oak_derive/tests/visit.rs, but I suppose it doesn't hurt to add a few here as well.

Before I do that though, maybe we should discuss if we actually want to try and eventually merge this PR into `main`? Do you think there is a need for supporting serde based formats, i.e. things like bincode, json etc., and do you think this is a good way to go about it?

Reply:
Yes I think as we move towards apps with more nodes, it is handy to have a way of specifying inter-node messages in Rust, without having to go all the way to protobuf, as long as they are only used internally. So something like this (or even simpler) would be super useful, and would help reduce the custom logic around serializing enums and structs, such as:

https://github.com/project-oak/oak/blob/1d84fbaebb689d8e858e941841c1147aaf68e381/oak_io/src/lib.rs#L52-L138